### PR TITLE
libovsdb: OperationResult fields optional

### PR DESCRIPTION
**Pull request by hunchback**

### CODE REVIEW

File: pkg/ovsdb/transact.go [hunk 1]
```diff
@@ -228,7 +228,10 @@ func NewTransaction(cli *clientv3.Client, request *libovsdb.Transact) *Transacti
 	txn.cache = Cache{}
 	txn.schemas = libovsdb.Schemas{}
 	txn.request = *request
-	txn.response.Result = make([]libovsdb.OperationResult, len(request.Operations))
+	txn.response.Result = make([]*libovsdb.OperationResult, len(request.Operations))
+	for i, _ := range txn.response.Result {
+		txn.response.Result[i] = libovsdb.NewOperationResult()
```
Comment (roytman) [inline pkg/ovsdb/transact.go:233]: This `NewOperationResult` call creates an instance of OperationResult which marshals  to JSON string as 
{"count":0,"error":"","details":"","uuid":["named-uuid",""],"rows":null}.
In order to be inline with the spec, I suggest just to create an empty OperationResult {}, and during the operations, just initialize the relevant field.

File: pkg/libovsdb/notation.go [hunk 2]
```diff
@@ -149,17 +149,27 @@ func NewTransact(params []interface{}) (*Transact, error) {
...
 // OperationResult is the result of an Operation
 type OperationResult struct {
-	Count   int         `json:"count,omitempty"`
-	Error   string      `json:"error,omitempty"`
-	Details string      `json:"details,omitempty"`
-	UUID    UUID        `json:"uuid,omitempty"`
-	Rows    []ResultRow `json:"rows,omitempty"`
+	Count   *int         `json:"count,omitempty"`
+	Error   *string      `json:"error,omitempty"`
+	Details *string      `json:"details,omitempty"`
```
Comment (roytman) [inline pkg/libovsdb/notation.go:160]: There is no `Details` filed

File: pkg/ovsdb/transact.go [hunk 3]
```diff
@@ -1020,8 +1023,8 @@ func doSelect(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.O
 		if err != nil {
 			return err
 		}
-		ovsResult.Count = ovsResult.Count + 1
-		ovsResult.Rows = append(ovsResult.Rows, *resultRow)
+		*ovsResult.Count++
```
Comment (roytman) [inline pkg/ovsdb/transact.go:1026]: before going into the for loop, we have to create the rows set.
*ovsResult.Rows = []libovsdb.Row{}
you don't need Count here

File: pkg/ovsdb/transact.go [hunk 4]
```diff
@@ -1045,11 +1048,11 @@ func doUpdate(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.O
...
-			ovsResult.Error = err.Error()
+			*ovsResult.Error = err.Error()
 			return err
 		}
 
-		ovsResult.Count = ovsResult.Count + 1
+		*ovsResult.Count++
 		err = RowUpdate(tableSchema, row, ovsOp.Row)
 		if err != nil {
 			return err
```
Comment (roytman) [inline pkg/ovsdb/transact.go:1058]: you just return errors, in each operation, but you have to set the Erro filed of the specific operation

File: pkg/ovsdb/transact.go [hunk 5]
```diff
@@ -1114,7 +1117,7 @@ func doDelete(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.O
 func preWait(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
 	if ovsOp.Timeout != 0 {
 		err := fmt.Errorf("only support timeout 0")
-		ovsResult.Error = err.Error()
+		*ovsResult.Error = err.Error()
```
Comment (roytman) [inline pkg/ovsdb/transact.go:1120]: this error set should be done for all operations

**Review by roytman — CHANGES_REQUESTED** [review]

File: pkg/libovsdb/client.go [hunk 6]
```diff
@@ -299,8 +299,8 @@ func (ovs OvsdbClient) MonitorCancel(jsonContext interface{}) error {
 	if err != nil {
 		return err
 	}
-	if reply.Error != "" {
-		return fmt.Errorf("Error while executing transaction: %s", reply.Error)
+	if *reply.Error != "" {
```
Comment (roytman) [inline pkg/libovsdb/client.go:302]: It's a part of the libovsdb, but we don't have to use OperationResult here, just return an error

File: pkg/ovsdb/transact.go [hunk 7]
```diff
@@ -980,7 +983,7 @@ func doInsert(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.O
 		for _, row := range txn.cache.Table(txn.request.DBName, ovsOp.Table) {
 			if (*row)[COL_UUIDNAME] == ovsOp.UUIDName {
```
Comment (roytman) [inline pkg/ovsdb/transact.go:984]: UUIDName can't be part of the cache, and the key, it is not UUID, and is provided only to  connect to different entries in the request .

Your logic here is wrong, you first, bring entire table, but when did not check the new UUID with the  cache

File: pkg/ovsdb/transact.go [hunk 8]
```diff
@@ -989,11 +992,11 @@ func doInsert(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.O
 	ok := txn.schemas.Validate(txn.request.DBName, ovsOp.Table, &ovsOp.Row)
 	if !ok {
 		err := errors.New(E_CONSTRAINT_VIOLATION)
-		ovsResult.Error = err.Error()
+		*ovsResult.Error = err.Error()
 		return err
 	}
 
-	ovsResult.Count = ovsResult.Count + 1
+	*ovsResult.Count++
```
Comment (roytman) [inline pkg/ovsdb/transact.go:999]: Insert doesn't return Count, only UUID or error

File: pkg/ovsdb/transact.go [hunk 9]
```diff
@@ -1020,8 +1023,8 @@ func doSelect(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.O
 		if err != nil {
```
Comment (roytman) [inline pkg/ovsdb/transact.go:1023]: 1. we have to set error in the operation result,
2.  we have to clean Row there

File: pkg/ovsdb/transact.go (hunk 3 above)
Comment (hunchback) [reply pkg/ovsdb/transact.go:1026]: ok

File: pkg/libovsdb/notation.go (hunk 2 above)
Comment (hunchback) [reply pkg/libovsdb/notation.go:160]: it was originally there and it is used in libovsdb/examples

File: pkg/ovsdb/transact.go (hunk 4 above)
Comment (hunchback) [reply pkg/ovsdb/transact.go:1053]: ok

File: pkg/ovsdb/transact.go (hunk 5 above)
Comment (hunchback) [reply pkg/ovsdb/transact.go]: ok

File: pkg/libovsdb/client.go (hunk 6 above)
Comment (hunchback) [reply pkg/libovsdb/client.go:302]: will leave as is

File: pkg/ovsdb/transact.go (hunk 8 above)
Comment (hunchback) [reply pkg/ovsdb/transact.go]: ok

File: pkg/ovsdb/transact.go (hunk 9 above)
Comment (hunchback) [reply pkg/ovsdb/transact.go:1022]: ok

**Pull request merged**